### ISSUE TITLE
join with engine=Join works only one time

### ISSUE THREAD

**Issue by den-crane** [ClickHouse/ClickHouse#16991]
```
20.11.1.5065.

CREATE TABLE test_join (serial String, appid UInt16) ENGINE = Join(ANY, INNER, serial);
INSERT INTO test_join (serial,appid) values ('aaa',123),('bbb',345);
CREATE TABLE test(serial String, val UInt32) ENGINE = TinyLog;
INSERT INTO test (serial,val) VALUES ('aaa',1111),('bbb',22222);


SELECT *
FROM test
ANY INNER JOIN test_join USING (serial)

Query id: 02c7efc9-0580-4292-ae1a-1bc98289c283

┌─serial─┬───val─┬─appid─┐
│ aaa    │  1111 │   123 │
│ bbb    │ 22222 │   345 │
└────────┴───────┴───────┘

2 rows in set. Elapsed: 0.001 sec.

:) select * from test any inner join test_join USING serial

SELECT *
FROM test
ANY INNER JOIN test_join USING (serial)

Query id: a3cd3230-9fd1-44d7-bf0b-da4b94e07081

Ok.

0 rows in set. Elapsed: 0.001 sec.

```

it worked in 20.3

**Comment by l1t1** [comment]
I tested, the rows in test_join is till there after 2nd Query returned empty
```sql
SELECT *
FROM test_join

┌─serial─┬─appid─┐
│ bbb    │   345 │
│ aaa    │   123 │
└────────┴───────┘
```

**Comment by l1t1** [comment]
and `ENGINE = Join(ALL, INNER, serial2);` is ok
```sql
CREATE TABLE test_join2 (serial2 String, appid UInt16) ENGINE = Join(ALL, INNER, serial2);
INSERT INTO test_join2(serial2,appid) VALUES ('aaa',1111),('bbb',22222);
SELECT * FROM test ALL INNER JOIN test_join2 on(test.serial= test_join2.serial2);

┌─serial─┬───val─┬─serial2─┬─appid─┐
│ aaa    │  1111 │ aaa     │  1111 │
│ bbb    │ 22222 │ bbb     │ 22222 │
└────────┴───────┴─────────┴───────┘

2 rows in set. Elapsed: 0.013 sec.

 SELECT * FROM test ALL INNER JOIN test_join2 on(test.serial= test_join2.serial2);


┌─serial─┬───val─┬─serial2─┬─appid─┐
│ aaa    │  1111 │ aaa     │  1111 │
│ bbb    │ 22222 │ bbb     │ 22222 │
└────────┴───────┴─────────┴───────┘

2 rows in set. Elapsed: 0.033 sec.

```

**Comment by filimonov** [comment]
>  alexey-milovidov removed the prio-major

@alexey-milovidov actually such non-deterministic behaviour doesn't sounds great, use case sounds valid, silent incorrect result is very bad behaviour, and older version worked. So all that sounds like the issue is important and require attention. 

So why not major?

**Comment by den-crane** [comment]
Maybe it's not major after all. Because Engine join supported only join on UInt64 keys before CH 20th something.

**Comment by l1t1** [comment]
ClickHouse client version 20.11.3.3 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.10.3 revision 54441.

the bug still exists

**Comment by filimonov** [comment]
> Maybe it's not major after all. Because Engine JOIN has supported only join on UInt64 keys before CH 20th something.

What do you mean? Your test perfectly works on 20.3, 19.11, 18.14.

**Comment by den-crane** [comment]
hm, maybe I am confused.

one more issue

https://github.com/ClickHouse/ClickHouse/blob/master/CHANGELOG.md

Add new ANY JOIN logic for StorageJoin consistent with JOIN operation. To upgrade without changes in behaviour you need add SETTINGS any_join_distinct_right_table_keys = 1 to Engine Join tables metadata or recreate these tables after upgrade. #8400 


```sql
set any_join_distinct_right_table_keys = 1;

CREATE TABLE test_join (serial String, appid UInt16) ENGINE = Join(ANY, INNER, serial);
INSERT INTO test_join (serial,appid) values ('aaa',123),('bbb',345);
CREATE TABLE test(serial String, val UInt32) ENGINE = TinyLog;
INSERT INTO test (serial,val) VALUES ('aaa',1111),('bbb',22222);

SELECT *
FROM test
ANY INNER JOIN test_join USING (serial)

Query id: e1456e28-79ce-4f0b-bb78-52c3beee155f


Received exception from server (version 20.12.1):
Code: 264. DB::Exception: Received from localhost:9000. DB::Exception: Table default.test_join has incompatible type of JOIN..

```

**Comment by vdimir** [comment]
> Add new ANY JOIN logic for StorageJoin consistent with JOIN operation. To upgrade without changes in behaviour you need add SETTINGS any_join_distinct_right_table_keys = 1 to Engine Join tables metadata or recreate these tables after upgrade. #8400

Cannot reproduce it with provided queries. Is there other conditions that leads to this failure?

**Comment by vdimir** [comment]
> and `ENGINE = Join(ALL, INNER, serial2);` is ok

Thanks for your notice it really helps. Actually it depend on type of join, bug was connected with `ANY INNER` join (and some others `RIGHT`, `FULL`).

But I see another problem here, we cannot use `JOIN ON (test.serial == test_join.serial)` an should rename it to serial2. I've created separate issue for it and will try to fix it #20309

**Comment by vdimir** [comment]
Issue from initial comment with StorageJoin that works only once fixed with PR #20079
